### PR TITLE
[App] Fixed Multi Node and add examples

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolves some bugs with the Multi Node components and add some example and tests.

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: tests/tests_app_examples/test_multi_node.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 import pytest
 from tests_app import _PROJECT_ROOT
 
-from lightning_app.testing.testing import application_testing, LightningTestApp
+from lightning.app.testing.testing import application_testing, LightningTestApp
```
Comment (lantiga) [inline tests/tests_app_examples/test_multi_node.py]: All examples use `lightning_app`, as imports are modified while packaging.
We need to be consistent throughout.

Since examples are likely to be copy-pasted, and imports at that point would be ran against the `lightning` package (and not `lightning_app`) then I think we should change this throughout.

I suggest to change all imports in a separate PR, and keep this one consistent (`lightning_app`).

File: examples/app_multi_node/app_pl_work.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import os
+
+import torch
+
+import lightning.app as L
```
Comment (lantiga) [inline examples/app_multi_node/app_pl_work.py:5]: I suggest we then modify all examples to `lightning.app` in a separate PR, since they will likely copy-pasted and ran from environments where the `lightning` package is installed.

**Review by lantiga — APPROVED** [review]
Look good, nice catch!

**Review by williamFalcon — CHANGES_REQUESTED** [review]
@tchaton 
Please standardize imports...

Wrong:
```python
import lightning.app as L

L.LightningApp
```

Right:
```python
import lightning as L

L.LightningApp
```

Wrong:
```python
from pytorch_lightning.callbacks import ModelCheckpoint
from pytorch_lightning.demos.boring_classes import BoringModel
```

Right
```python
from lightning.pytorch.callbacks import ModelCheckpoint
from lightning.pytorch.demos.boring_classes import BoringModel
```

File: tests/tests_app_examples/test_multi_node.py (hunk 1 above)
Comment (williamFalcon) [reply tests/tests_app_examples/test_multi_node.py]: @lantiga externally, we use lightning.app not lightning_app.

We need a consistent experience for users. Having lightning_app and lightning.app all over the place is already confusing people.

File: examples/app_multi_node/app_pl_work.py (hunk 2 above)
Comment (williamFalcon) [reply examples/app_multi_node/app_pl_work.py:5]:
```suggestion
import lightning as L
```

File: examples/app_multi_node/app_pl_work.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from lightning.app.components import MultiNode
+from pytorch_lightning import Trainer
```
Comment (williamFalcon) [inline examples/app_multi_node/app_pl_work.py:7]:
```suggestion

```

File: examples/app_multi_node/app_pl_work.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        nodes: int,
+        node_rank: int,
+    ):
+        os.environ["MASTER_ADDR"] = main_address
+        os.environ["MASTER_PORT"] = str(main_port)
+        os.environ["NODE_RANK"] = str(node_rank)
+
+        devices = torch.cuda.device_count() if torch.cuda.is_available() else 2
+        model = LoggingBoringModel()
+        trainer = Trainer(
```
Comment (williamFalcon) [inline examples/app_multi_node/app_pl_work.py:33]:
```suggestion
        trainer = L.Trainer(
```

File: examples/app_multi_node/app_pl_work.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import os
+
+import torch
+
+import lightning.app as L
+from lightning.app.components import MultiNode
+from pytorch_lightning import Trainer
+from pytorch_lightning.callbacks import ModelCheckpoint
```
Comment (williamFalcon) [inline examples/app_multi_node/app_pl_work.py:8]:
```suggestion
from lightning.pytorch.callbacks import ModelCheckpoint
```

File: examples/app_multi_node/app_pl_work.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+from pytorch_lightning.demos.boring_classes import BoringModel
```
Comment (williamFalcon) [inline examples/app_multi_node/app_pl_work.py:9]:
```suggestion
from lightning.pytorch.demos.boring_classes import BoringModel
```

**Review by williamFalcon — CHANGES_REQUESTED** [review]

File: examples/app_multi_node/app_torch_work.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
+import torch
+from torch.nn.parallel.distributed import DistributedDataParallel
+
+import lightning.app as L
```
Comment (williamFalcon) [inline examples/app_multi_node/app_torch_work.py:4]:
```suggestion
import lightning as L
```

File: tests/tests_app_examples/test_multi_node.py (hunk 1 above)
Comment (tchaton) [reply tests/tests_app_examples/test_multi_node.py]: @lantiga This is changed only during debugging.

File: examples/app_multi_node/app_pl_work.py (hunk 4 above)
Comment (tchaton) [reply examples/app_multi_node/app_pl_work.py]: Done !

File: examples/app_multi_node/app_pl_work.py (hunk 5 above)
Comment (tchaton) [reply examples/app_multi_node/app_pl_work.py]: Done !

File: examples/app_multi_node/app_pl_work.py (hunk 6 above)
Comment (tchaton) [reply examples/app_multi_node/app_pl_work.py]: Done !

File: examples/app_multi_node/app_torch_work.py (hunk 7 above)
Comment (tchaton) [reply examples/app_multi_node/app_torch_work.py]: Done !

**Review by williamFalcon — APPROVED** [review]

File: examples/app_multi_node/app_pl_work.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import os
+
+import lightning as L
+from lightning.app.components import MultiNode
+from lightning.pytorch.demos.boring_classes import BoringModel
+
+
+class PyTorchLightningMultiNode(L.LightningWork):
```
Comment (lantiga) [inline examples/app_multi_node/app_pl_work.py:8]: In order to reduce confusion, we could name this `PyTorchLightningDistributed`.

This way conceptually you pass a `Distributed`-aware PyTorchLightning script to a `MultiNode` orchestrator. Otherwise IMO it's kind of confusing to be passing a `MultiNode` something to another `MultiNode` something

File: examples/app_multi_node/app_torch_work.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    # 5. Verify all processes have the same weights at the end of training.
+    weight = model.module.weight.clone()
+    torch.distributed.all_reduce(weight)
+    assert torch.equal(model.module.weight, weight / world_size)
+
+    print("Multi Node Distributed Training Done!")
+
+
+class PyTorchMultiNode(L.LightningWork):
```
Comment (lantiga) [inline examples/app_multi_node/app_torch_work.py:49]: Same thing, this could be named `PyTorchDistributed`

**Review by lantiga — APPROVED** [review]
Looks good. I have a naming suggestion that IMO makes things a bit clearer.

File: examples/app_multi_node/app_torch_work.py (hunk 9 above)
Comment (tchaton) [reply examples/app_multi_node/app_torch_work.py:49]: Done !

File: examples/app_multi_node/app_pl_work.py (hunk 8 above)
Comment (tchaton) [reply examples/app_multi_node/app_pl_work.py]: Done !

**Review by lantiga — APPROVED** [review]

File: src/lightning_app/components/multi_node.py [hunk 10]
```diff
@@ -90,6 +90,7 @@ def run(self) -> None:
             self.ws[node_rank].run(
                 main_address=self.ws[0].internal_ip,
                 main_port=self.ws[0].port,
+                nodes=self.nodes,
```
Comment (rohitgr7) [inline src/lightning_app/components/multi_node.py:93]: I think a better name would be `num_nodes`.
Also need to update the example in the docstring: `AnyDistributedComponent.run`

File: examples/app_multi_node/app_torch_work.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        output = model(x)
+        loss = criterion(output, torch.ones_like(output))
+        print(f"global_rank: {global_rank} step: {step} loss: {loss}")
+        loss.backward()
+        optimizer.step()
+
+    # 5. Verify all processes have the same weights at the end of training.
+    weight = model.module.weight.clone()
+    torch.distributed.all_reduce(weight)
+    assert torch.equal(model.module.weight, weight / world_size)
```
Comment (rohitgr7) [inline examples/app_multi_node/app_torch_work.py:44]:
```suggestion
    assert torch.equal(model.module.weight, weight)
```
we are doing all_reduce, taking average?

**Review by rohitgr7 — APPROVED** [review]

File: examples/app_multi_node/app_pl_work.py (hunk 8 above)
Comment (rohitgr7) [reply examples/app_multi_node/app_pl_work.py]: how about naming it `LightningTrainerDistributed`?

File: examples/app_multi_node/app_torch_work.py (hunk 11 above)
Comment (tchaton) [reply examples/app_multi_node/app_torch_work.py:44]: It is sum by default. So you need to divide by world_size.

File: src/lightning_app/components/multi_node.py (hunk 10 above)
Comment (tchaton) [reply src/lightning_app/components/multi_node.py:93]: Done !

File: examples/app_multi_node/app_pl_work.py (hunk 8 above)
Comment (tchaton) [reply examples/app_multi_node/app_pl_work.py]: I prefer Luca proposition :)

**Review by lantiga — APPROVED** [review]

**Review by manskx — APPROVED** [review]
